chainer/training/triggers/interval.py

```diff
@@ -40,9 +42,16 @@ def __call__(self, trainer):
         """
         updater = trainer.updater
         if self.unit == 'epoch':
-            prev = self.count
-            self.count = updater.epoch_detail // self.period
-            return prev != self.count
+            epoch_detail = updater.epoch_detail
+            previous_epoch_detail = updater.previous_epoch_detail
```

Comment:
Hi @Hakuyume

Sorry for our late response. I will share you the discussion as I promised. Although, we have found some negative points of the idea of controlling `IntervalTrigger` with  `Updater.previous_epoch_detail`.

IntervalTrigger cannot rely on Updater.previous_epoch_detail if it is **not called at every epoch**.
Suppose we have a instance of IntervalTrigger whose interval is 1 epoch and 1 epoch consists of 4 iterations. Consider the following situation:

```
 1  2   3   4    5   6   7   8   9
---|---|---|---||---|---|---|---||---|
 x       x              o       x
```

The numbers are iteration counts and x and o indicate the iteration the trigger is *called* (not *fired*. The trigger is fired only when `trigger(trainer)` is `True`). We want to pay attention to the 6th iteration (indicated as o). As this is the first iteration of the second epoch, the trigger should be fired. But, as updater.previous_epoch_detail is 1.25 (=5/4), updater.previous_epoch_detail is 1.5, the trigger cannot fire itself. In order to make the trigger fire at the iteration, the trigger must memorize the last time it is called by ourselves, otherwise, trainers are responsible for managing the last time each trigger is called.

We encountered this problem when we use `IntervalTrigger`s to collect the statistics of data nodes in the computational graph. We associate a trigger for each data node in the computational graph. But because of the define-by-run nature of Chainer, not all nodes appeared at every forward propagation (we somehow associate data nodes in the computational graphs of different iterations somehow).

If we assume that triggers are called every iteration, this problem does not happen. Actually it is true of triggers for controlling the invocation of extensions, that are fed with `trigger` option of the `Trainer.extend` method. But this assumption restricts the usage of triggers. Actually one of our members find it when he uses triggers to collect the statistics of Chainer variables.

At first, I thought it is somewhat strange the `IntervalTrigger` is stateful. But we found that we already have stateful triggers (e.g. `BestValueTrigger`). So, we need to serialize triggers anyway.

Of course this it is not the final decision, we are open to discussion with you and of course anyone who is interested in this topic. I would like to know your opinion and I promise I will reply to you as quickly as possible in the next time.